src/qibo/models/encodings.py

```diff
@@ -1787,3 +1974,213 @@ def _add_wbd_gate(
+                        continue
+                    # There is a typo in the paper
+                    # b_{j},{kprime} should be different from Ã_{j},{kprime} not Ã_{j},{k}
+                    if A[j, kprime] == atilde[j, kprime]:
+                        continue
+
+                    qgates.append(gates.CNOT(n - k - 1, n - kprime - 1))
+                    # check whether the gate is applied on other rows
+                    for l in range(nrows):
+                        if A[l, k] == 1:
+                            A[l, kprime] = (A[l, kprime] + 1) % 2
+
+                # Let us clean the element b_{j},{k}
+
+                # There is another typo in the paper
+                # the control qubits for this gate correspond to the non-zero elements in row j of matrix A, not Ã
+                ctrls = [n - l - 1 for l in range(k) if A[j, l] != 0]
+                qgates.append(gates.X(n - k - 1).controlled_by(*ctrls))
+                ctrls = [l for l in range(k) if A[j, l] != 0]
+
+                # check whether the gate is applied on other rows
+                for l in range(nrows):
+                    if backend.np.array_equal(
+                        A[l, ctrls], backend.np.ones_like(A[l, ctrls])
+                    ):
+                        A[l, k] = (A[l, k] + 1) % 2
+
+    return qgates
+
+
+def _perm_pair_flip_ops(n: int, m: int, backend=None):
+    """Implement σ_{i,2} as X fan‑in + MCX + X fan‑out."""
+    backend = _check_backend(backend)
+    # let us flip the first qubit when the last {int(n-math.log2(2*m))} qubits are all in the state |0⟩
+    prefix = int(backend.np.ceil(backend.np.log2(2 * m)))
+    x_qubits = list(range(prefix, n))
+    qgates = [gates.X(n - q - 1) for q in x_qubits]
+    qgates.append(
+        gates.X(n - 1).controlled_by(*range(n - int(np.ceil(math.log2(2 * m)))))
+    )  # flip qubit 0
+    qgates.extend(gates.X(n - q - 1) for q in x_qubits)
+
+    return qgates
+
+
+def permutation_synthesis(
+    sigma: list[int],
+    m: int = 2,
+    backend=None,
+):
+    """Given permutation sigma on {0,...,d-1} and a power‑of‑two budget m,
+    factor sigma into the fewest layers sigma_1, sigma_2, ..., sigma_t such that
+        - each layer has at most m disjoint transpositions
+        - each layer moves a power‑of‑two number of indices.
+        Return a circuit synthesis of sigma.
+
+    Args:
+        sigma (list[int]): permutation description on {0,...,d-1}.
+        m (int): power‑of‑two budget m (Default:2)
+
+    Returns:
+        :class:`qibo.models.circuit.Circuit`: Circuit that implements the permutation sigma.
+
+    References:
+        1. L. Li, J. Luo,
+        *Nearly Optimal Circuit Size for Sparse Quantum State Preparation*
+        `arXiv:2406.16142 (2024) <https://doi.org/10.48550/arXiv.2406.16142>`_.
+    """
+    backend = _check_backend(backend)
+
+    if not isinstance(sigma, list):
+        raise_error(TypeError, f"Permutation sigma must be list[int]")
```

Comment:
```suggestion
    if not isinstance(sigma, list):
        raise_error(TypeError, f"Permutation ``sigma`` must be a ``list`` of ``int``.")
```

I guess this could also be a ``tuple`` or an ``ndarray``, just to be more general.